### ISSUE TITLE
Spec.__contains__ produces wrong results for `intel-oneapi-mkl`

### ISSUE THREAD

**Issue by alalazo** [LLNL/spack#35597]
### Steps to reproduce

Reported by @dev-zero in #35584

```python
Spack version 0.20.0.dev0
Python 3.8.10, Linux x86_64
>>> from spack.spec import Spec
>>> s = Spec("cp2k ^intel-oneapi-mkl")
>>> c = s.concretized()
>>> "intel-oneapi-mkl" in c
True
>>> "^intel-oneapi-mkl" in c
False
>>> c["blas"]
intel-oneapi-mkl@2023.0.0%gcc@9.4.0~cluster+envmods~ilp64+shared build_system=generic arch=linux-ubuntu20.04-icelake ^intel-oneapi-tbb@2021.8.0%gcc@9.4.0+envmods build_system=generic arch=linux-ubuntu20.04-icelake
>>> c["lapack"]
intel-oneapi-mkl@2023.0.0%gcc@9.4.0~cluster+envmods~ilp64+shared build_system=generic arch=linux-ubuntu20.04-icelake ^intel-oneapi-tbb@2021.8.0%gcc@9.4.0+envmods build_system=generic arch=linux-ubuntu20.04-icelake
```

### Error message

The entry:
```python
>>> "^intel-oneapi-mkl" in c
False
```
should return `True`.

### Information on your system

* **Spack:** 0.20.0.dev0 (fbd5c5c2cdc6761cb6cd5dfd2d0c33fdfcdb1712)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
It seems for the `intel-oneapi-mkl` case we do something wrong here:

https://github.com/spack/spack/blob/5b8917188a4d84713c1037fd9464bf77503edd27/lib/spack/spack/spec.py#L3574-L3584

**Comment by alalazo** [comment]
I think we should have an answer before hitting the lines above, since `"^intel-oneapi-mkl" in c` doesn't involve any virtual dependency at all :thinking:

**Comment by alalazo** [comment]
Trying to investigate if this:
```diff
diff --git a/lib/spack/spack/spec.py b/lib/spack/spack/spec.py
index 94e3230285..3b9019807e 100644
--- a/lib/spack/spack/spec.py
+++ b/lib/spack/spack/spec.py
@@ -3571,6 +3571,9 @@ def satisfies_dependencies(self, other, strict=False):
             if not self[name].satisfies(other[name], deps=False):
                 return False
 
+        if not any(x.virtual for x in other.traverse()):
+            return True
+
         # For virtual dependencies, we need to dig a little deeper.
         self_index = spack.provider_index.ProviderIndex(
             repository=spack.repo.path, specs=self.traverse(), restrict=True
```
might be a good lead to solve the issue.

**Comment by alalazo** [comment]
Fixed in #35681
